### ISSUE TITLE
The language server does not trigger dialyzer

### ISSUE THREAD

**Issue by sashaafm** [elixir-lsp/elixir-ls#295]
### Problem
Just yesterday I updated a lot of stuff on my machine (all `brew` packages, MacOS itself, vim plugins) and I downloaded the most recent ElixirLS version. After this, I've got a certain project where I can't seem to be able to trigger dialyzer to run when ElixirLS starts up. This was working fine before. I'm using CoC and it works for my other language server (Elm language server, for example). Here's the output:

```
## versions

vim version: VIM - Vi IMproved 8.2 8020950
node version: v14.4.0
coc.nvim version: 0.0.78
coc.nivm directory: /Users/sashaafm/.vim/bundle/coc.nvim
term: iTerm.app
platform: darwin

## Output channel: languageserver.elixirLS

Started ElixirLS v0.4.0
Elixir version: "1.10.2 (compiled with Erlang/OTP 21)"
Erlang version: "22"
Compiling with Mix env test
[Info  - 12:05:17 PM] Compile took 73 milliseconds
```

I believe after this line it was supposed to start emitting lines about Dialyzer and Symbols. I've tried the most common troubleshooting patterns and this wasn't solved.

### Environment

* Elixir & Erlang versions (elixir --version): 
```
Erlang/OTP 22 [erts-10.7.2] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [hipe]

Elixir 1.10.2 (compiled with Erlang/OTP 21)
```
* Operating system: 
`Darwin Zane 19.5.0 Darwin Kernel Version 19.5.0: Tue May 26 20:41:44 PDT 2020; root:xnu-6153.121.2~2/RELEASE_X86_64 x86_64` (MacOS 10.15.5)
* Editor or IDE name (e.g. Emacs/VSCode): 
```
VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Jun 10 2020 16:42:10)
macOS version
Included patches: 1-950
```

* LSP Client name: 
ElixirLS 0.4.0

### Troubleshooting

- [X] Restart your editor (which will restart ElixirLS) sometimes fixes issues
- [X] Stop your editor, remove the entire `.elixir_ls` directory, then restart your editor
  * NOTE: This will cause you to have to re-run the dialyzer build for your project

### Logs

2.  Create a new Mix project with `mix new empty`, then open that project with VS Code and open an Elixir file. Is your issue reproducible on the empty project? If not, please publish a repo on Github that does reproduce it.
I tried this and it's correctly working, I think. The output shows this:
```
Started ElixirLS v0.4.0
Elixir version: "1.10.2 (compiled with Erlang/OTP 21)"
Erlang version: "22"
Compiling with Mix env test
Compiling 1 file (.ex)
Generated empty app
[Info  - 12:03:02 PM] Compile took 1634 milliseconds
[Info  - 12:03:02 PM] [ElixirLS Dialyzer] Checking for stale beam files
[Info  - 12:03:02 PM] [ElixirLS Dialyzer] Found 7 changed files in 5 milliseconds
[Info  - 12:03:03 PM] [ElixirLS Dialyzer] Analyzing 7 modules: [Collectable, Empty, Enumerable, IEx.Info, Inspect, List.Chars, String.Chars]
[Info  - 12:03:03 PM] [ElixirLS Dialyzer] Analysis finished in 976 milliseconds
[Info  - 12:03:03 PM] Dialyzer analysis is up to date
[Info  - 12:03:04 PM] [ElixirLS Dialyzer] Writing manifest...
[Info  - 12:03:04 PM] [ElixirLS Dialyzer] Done writing manifest in 1154 milliseconds.
[Info  - 12:03:05 PM] [ElixirLS WorkspaceSymbols] Indexing...
[Info  - 12:03:11 PM] [ElixirLS WorkspaceSymbols] Module discovery complete
[Info  - 12:03:11 PM] [ElixirLS WorkspaceSymbols] 25 callbacks added to index
[Info  - 12:03:11 PM] [ElixirLS WorkspaceSymbols] 299 modules added to index
[Info  - 12:03:12 PM] [ElixirLS WorkspaceSymbols] 488 types added to index
[Info  - 12:03:15 PM] [ElixirLS WorkspaceSymbols] 6218 functions added to index
```

I cannot publish a repo with the problem due to confidentiality of the code, and it being quite a large project.

EDIT:

I noticed a line in my Vim, but didn't manage to copy it. It was something about "changeNotifications" "after 5 seconds" and started using the default settings

EDIT 2:

Somehow managed to get this error on my output:

```
[Warn  - 12:44:50 PM] warning: Received unmatched notification: %{"jsonrpc" => "2.0", "method" => "workspace/didChangeWorkspaceFolders", "params" => %{"event" => %{"added" => [%{"name" => "web", "uri" => "file:///Users/sashaafm/Documents/project/frontend/web"}], "removed" => []}}}
  (language_server 0.4.0) lib/language_server/server.ex:321: ElixirLS.LanguageServer.Server.handle_notification/2
  (language_server 0.4.0) lib/language_server/server.ex:134: ElixirLS.LanguageServer.Server.handle_cast/2
  (stdlib 3.12.1) gen_server.erl:637: :gen_server.try_dispatch/4
  (stdlib 3.12.1) gen_server.erl:711: :gen_server.handle_msg/6
  (stdlib 3.12.1) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
```

EDIT 3:
I've managed to trigger some further errors

```
Started ElixirLS v0.4.0
Elixir version: "1.10.2 (compiled with Erlang/OTP 21)"
Erlang version: "22"
Compiling with Mix env test
[Info  - 2:26:50 PM] Compile took 76 milliseconds
/Users/sashaafm/.asdf/installs/erlang/22.3.4/lib/compiler-7.5.4/src/compile.erl: 
at line 39: 
can't find include file "erl_compile.hrl"

[Warn  - 2:28:06 PM] docsh error: error
[{edoc,'-check_forms/3-fun-0-',3,[{file,"edoc.erl"},{line,712}]},
 {lists,foreach,2,[{file,"lists.erl"},{line,1338}]},
 {edoc,read_source,2,[{file,"edoc.erl"},{line,581}]},
 {edoc_extract,source,3,[{file,"edoc_extract.erl"},{line,57}]},
 {docsh_edoc,to_internal,2,
             [{file,"/home/jason/dev/forks/elixir-ls/deps/docsh/src/docsh_edoc.erl"},
              {line,43}]},
 {docsh_lib,convert_one,1,
            [{file,"/home/jason/dev/forks/elixir-ls/deps/docsh/src/docsh_lib.erl"},
             {line,42}]},
 {lists,flatmap,2,[{file,"lists.erl"},{line,1250}]},
 {docsh_lib,convert,3,
            [{file,"/home/jason/dev/forks/elixir-ls/deps/docsh/src/docsh_lib.erl"},
             {line,37}]}]

Compiling with Mix env test
[Info  - 2:28:08 PM] Compile took 0 milliseconds
Compiling with Mix env test
[Info  - 2:28:25 PM] Compile took 0 milliseconds
Compiling with Mix env test
[Info  - 2:31:07 PM] Compile took 0 milliseconds
Compiling with Mix env test
[Info  - 2:31:10 PM] Compile took 0 milliseconds
Compiling with Mix env test
[Info  - 2:31:23 PM] Compile took 0 milliseconds
Compiling with Mix env test
[Info  - 2:31:46 PM] Compile took 0 milliseconds
/Users/sashaafm/.asdf/installs/erlang/22.3.4/lib/kernel-6.5.2/src/application_controller.erl: 
at line 47: 
can't find include file "logger.hrl"

[Warn  - 2:32:09 PM] docsh error: error
[{edoc,'-check_forms/3-fun-0-',3,[{file,"edoc.erl"},{line,712}]},
 {lists,foreach,2,[{file,"lists.erl"},{line,1338}]},
 {edoc,read_source,2,[{file,"edoc.erl"},{line,581}]},
 {edoc_extract,source,3,[{file,"edoc_extract.erl"},{line,57}]},
 {docsh_edoc,to_internal,2,
             [{file,"/home/jason/dev/forks/elixir-ls/deps/docsh/src/docsh_edoc.erl"},
              {line,43}]},
 {docsh_lib,convert_one,1,
            [{file,"/home/jason/dev/forks/elixir-ls/deps/docsh/src/docsh_lib.erl"},
             {line,42}]},
 {lists,flatmap,2,[{file,"lists.erl"},{line,1250}]},
 {docsh_lib,convert,3,
            [{file,"/home/jason/dev/forks/elixir-ls/deps/docsh/src/docsh_lib.erl"},
             {line,37}]}]

Compiling with Mix env test
[Info  - 2:32:16 PM] Compile took 0 milliseconds
```

Maybe this is related to `asdf` usage?

EDIT 4:

Uninstalled `asdf` and now I'm getting this. Also a `beam.smp` process is now taking all my CPU running a `ElixirLanguageServer.CLI.main()` function which I think is supposed to happen, right?
```
Started ElixirLS v0.4.0
Elixir version: "1.10.3 (compiled with Erlang/OTP 22)"
Erlang version: "23"
[Error  - 2:43:50 PM] Request textDocument/completion failed.
  Message: an exception was raised:
    ** (MatchError) no match of right hand side value: []
        (elixir_sense 1.0.1) lib/elixir_sense/core/state.ex:601: ElixirSense.Core.State.remove_vars_scope/1
        (elixir_sense 1.0.1) lib/elixir_sense/core/metadata_builder.ex:100: ElixirSense.Core.MetadataBuilder.post_module/3
        (elixir_sense 1.0.1) lib/elixir_sense/core/metadata_builder.ex:46: ElixirSense.Core.MetadataBuilder.build/1
        (elixir_sense 1.0.1) lib/elixir_sense/core/parser.ex:30: ElixirSense.Core.Parser.parse_string/4
        (language_server 0.4.0) lib/language_server/providers/completion.ex:101: ElixirLS.LanguageServer.Providers.Completion.completion/4
        (language_server 0.4.0) lib/language_server/server.ex:538: anonymous fn/3 in ElixirLS.LanguageServer.Server.handle_request_async/2
  Code: -32000 

14:43:50.976 [error] Process #PID<0.120.0> raised an exception
** (MatchError) no match of right hand side value: []
    (elixir_sense 1.0.1) lib/elixir_sense/core/state.ex:601: ElixirSense.Core.State.remove_vars_scope/1
    (elixir_sense 1.0.1) lib/elixir_sense/core/metadata_builder.ex:100: ElixirSense.Core.MetadataBuilder.post_module/3
    (elixir_sense 1.0.1) lib/elixir_sense/core/metadata_builder.ex:46: ElixirSense.Core.MetadataBuilder.build/1
    (elixir_sense 1.0.1) lib/elixir_sense/core/parser.ex:30: ElixirSense.Core.Parser.parse_string/4
    (language_server 0.4.0) lib/language_server/providers/completion.ex:101: ElixirLS.LanguageServer.Providers.Completion.completion/4
    (language_server 0.4.0) lib/language_server/server.ex:538: anonymous fn/3 in ElixirLS.LanguageServer.Server.handle_request_async/2
Compiling with Mix env test
[Info  - 2:43:55 PM] Compile took 11 milliseconds
Compiling with Mix env test
[Info  - 2:44:48 PM] Compile took 0 milliseconds
Compiling with Mix env test
[Info  - 2:45:02 PM] Compile took 0 milliseconds
Compiling with Mix env test
[Info  - 2:45:25 PM] Compile took 0 milliseconds
Compiling with Mix env test
[Info  - 2:45:37 PM] Compile took 0 milliseconds
```

EDIT 5:
Looks like it successfully built the PLT for the Elixir/Erlang version. Now it was supposed to build the PLT for my project, I think? Can't seem to trigger that

EDIT 6:
This will be my last edit for now while I wait for some support. I added the flag for more verbose output and got this:
```
[Trace - 3:12:38 PM] Received notification 'textDocument/publishDiagnostics'.
Params: {
    "diagnostics": [
        {
            "message": "No mixfile found in project. To use a subdirectory, set `elixirLS.projectDir` in your settings",
            "range": {
                "end": {
                    "character": 0,
                    "line": 0
                },
                "start": {
                    "character": 0,
                    "line": 0
                }
            },
            "severity": 1,
            "source": "ElixirLS"
        }
    ],
    "uri": "file:///Users/sashaafm/Documents/project/mix.exs"
}
```

**Comment by axelson** [comment]
@sashaafm What does the project layout of your project look like? It looks like your `mix.exs` is not in the root of your project. Also I filed your `unmatched notification` warning as #296, fixing that may slightly improve the situation. Do you have multiple workspaces (or some vim equivalent)?

**Comment by sashaafm** [comment]
Hello @axelson my project layout is like this `overall_project_root_dir/dir_1/dir_2/elixir_project_root_dir`

This being that `overall_project_root_dir` is where the Git repository was initiated. Then there are two intermediate directories `dir_1` and `dir_2` and finally the Elixir project directory where the `mix.exs` file resides (this is an umbrella project).

As I had said this was working perfectly before I updated my machine's overall software and when I updated to the latest `0.4.0` version. Did something change in this version regarding project layouts?

Thank you

**Comment by axelson** [comment]
@sashaafm do you the specific version of ElixirLS you were using previously? Nothing about detecting the location of mix.exs location has changed recently. And have you tried setting the `projectDir` in the language server settings? (You'll have to set it through coc.vim somehow).

**Comment by sashaafm** [comment]
@axelson I tried settings `projectDir` by creating a local CoC settings file via `:CocLocalConfig` but that doesn't seem to work also.

I'm not sure what the previous version was. I believe it was `0.3.3` because I only started using CoC around May.

**Comment by sashaafm** [comment]
@axelson I tried deleting all my Vim plugins as well as CoC config folder and installing everything again. Still not working. I tried tracing again and got some further info:
```
[Trace - 11:55:49 AM] Sending notification 'textDocument/didOpen'.
Params: {
    "textDocument": {
        "uri": "file:///Users/sashaafm/Documents/project/dir_1/dir_2/mix.exs",
        "languageId": "elixir",
        "version": 1,
        "text": <REDACTED>
    }
}

[Trace - 11:55:49 AM] Received notification 'textDocument/publishDiagnostics'.
Params: {
    "diagnostics": [],
    "uri": "file:///Users/sashaafm/Documents/project/dir_1/dir_2/mix.exs"
}


[Trace - 11:55:49 AM] Received request 'client/registerCapability - (1)'.
Params: {
    "registrations": [
        {
            "id": "B63DDB275E15ED256156A5090CE35EB12FC8CC8C",
            "method": "workspace/didChangeWatchedFiles",
            "registerOptions": {
                "watchers": [
                    {
                        "globPattern": "**/*.ex"
                    },
                    {
                        "globPattern": "**/*.exs"
                    },
                    {
                        "globPattern": "**/*.eex"
                    },
                    {
                        "globPattern": "**/*.leex"
                    }
                ]
            }
        }
    ]
}


[Trace - 11:55:49 AM] Sending response 'client/registerCapability - (1)'. Processing request took 4ms
No result returned.
```

Is this any help?

**Comment by sashaafm** [comment]
@axelson after installing the `master` branch version I still get this error:
```
Params: {
    "diagnostics": [
        {
            "message": "No mixfile found in project. To use a subdirectory, set `elixirLS.projectDir` in your settings",
            "range": {
                "end": {
                    "character": 0,
                    "line": 0
                },
                "start": {
                    "character": 0,
                    "line": 0
                }
            },
            "severity": 1,
            "source": "ElixirLS"
        }
    ],
    "uri": "file:///Users/sashaafm/Documents/project/mix.exs"
}
```

I believe this is the root cause of the problem. However, I've already tried adding the `projectDir` option in the config of CoC and it doesn't seem to be recognised? Is this something from LSP that must be implemented by `elixir-lsp` package?

I'm at a loss trying to fix this. The language server is working perfectly on my other projects except for this one.

**Comment by axelson** [comment]
I'm not familiar with setting configuration from vim/CoC but can you check that you're setting `elixirLS.projectDir` rather than a plain `projectDir`?

**Comment by sashaafm** [comment]
Hello yet again @axelson, as I had said in the Slack channel I've finally managed to get `elixir-lsp` working again. 

My solution was to create a local configuration file for CoC using the `:CocLocalConfig` command and I just copied my `elixirLS` config from the global config (in `:CocConfig`). Then deleted the `.elixir_ls` directory and started vim in one of the project's source code files and it finally started working correctly again.

I'm not sure this is the ideal solution to this problem, but at least it works. I think ideally it would correctly use `elixirLS.projectDir` configuration key, but I couldn't get this key to work even when creating a local configuration file.
